src/server/worker.cc

```diff
@@ -442,115 +439,146 @@ Status Worker::EnableWriteEvent(int fd) {
 
 Status Worker::Reply(int fd, const std::string &reply) {
-  std::unique_lock<std::mutex> lock(conns_mu_);
-  auto iter = conns_.find(fd);
-  if (iter != conns_.end()) {
-    iter->second->SetLastInteraction();
-    redis::Reply(iter->second->Output(), reply);
+  if (ConnMap::accessor accessor; conns_.find(accessor, fd)) {
+    accessor->second->SetLastInteraction();
+    redis::Reply(accessor->second->Output(), reply);
     return Status::OK();
   }
 
   return {Status::NotOK, "connection doesn't exist"};
 }
 
 void Worker::BecomeMonitorConn(redis::Connection *conn) {
-  {
-    std::lock_guard<std::mutex> guard(conns_mu_);
-    conns_.erase(conn->GetFD());
-    monitor_conns_[conn->GetFD()] = conn;
+  if (ConnMap::accessor accessor; conns_.find(accessor, conn->GetFD())) {
+    conns_.erase(accessor);
+    accessor.release();
+
+    if (ConnMap::accessor accessor; monitor_conns_.find(accessor, conn->GetFD())) {
+      accessor->second = conn;
+    } else {
+      monitor_conns_.insert(accessor, std::make_pair(conn->GetFD(), conn));
+    }
   }
   srv->IncrMonitorClientNum();
   conn->EnableFlag(redis::Connection::kMonitor);
 }
 
 void Worker::QuitMonitorConn(redis::Connection *conn) {
-  {
-    std::lock_guard<std::mutex> guard(conns_mu_);
-    monitor_conns_.erase(conn->GetFD());
-    conns_[conn->GetFD()] = conn;
+  if (ConnMap::accessor accessor; monitor_conns_.find(accessor, conn->GetFD())) {
+    {
+      monitor_conns_.erase(accessor);
+      accessor.release();
+    }
+    if (ConnMap::accessor accessor; conns_.find(accessor, conn->GetFD())) {
+      accessor->second = conn;
+    } else {
+      conns_.insert(accessor, std::make_pair(conn->GetFD(), conn));
+    }
   }
   srv->DecrMonitorClientNum();
   conn->DisableFlag(redis::Connection::kMonitor);
 }
 
 void Worker::FeedMonitorConns(redis::Connection *conn, const std::string &response) {
-  std::unique_lock<std::mutex> lock(conns_mu_);
-
-  for (const auto &iter : monitor_conns_) {
-    if (conn == iter.second) continue;  // skip the monitor command
-
-    if (conn->GetNamespace() == iter.second->GetNamespace() || iter.second->GetNamespace() == kDefaultNamespace) {
-      iter.second->Reply(response);
+  tbb::parallel_for(monitor_conns_.range(), [conn, response](const ConnMap::range_type &range) {
+    for (auto &it : range) {
+      const auto &value = it.second;
+      if (conn == value) continue;  // skip the monitor command
+      if (conn->GetNamespace() == value->GetNamespace() || value->GetNamespace() == kDefaultNamespace) {
+        value->Reply(response);
+      }
     }
-  }
+  });
 }
 
 std::string Worker::GetClientsStr() {
-  std::unique_lock<std::mutex> lock(conns_mu_);
-
-  std::string output;
-  for (const auto &iter : conns_) {
-    redis::Connection *conn = iter.second;
-    output.append(conn->ToString());
-  }
-
-  return output;
+  return tbb::parallel_reduce(
+      conns_.range(), std::string{},
+      [](const ConnMap::range_type &range, std::string &&result) {
```

Comment:
Emm why `&&result`? Using `std::string result` here?

Reply:
Hi @mapleFU ,

Previously I thought the accumulation result should be a temporary value, so I used this signature.
But this right value reference is not necessary here.
If it's a returned temporary value, the RVO should work to help optimize.
I have updated it with a normal argument copy type.

Best Regards,
Edward